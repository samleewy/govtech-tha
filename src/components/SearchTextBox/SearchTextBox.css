input[type=text] {
    color: #282828;
    border: 1px solid #A4A4A4;
    border-radius: 8px;
    padding:12px;
    width: 100%;
}

input[type=text]:hover {
    border:1px solid #1C76D5;
}

input[type=text]:focus {
    outline:none;
    border:1px solid #1C76D5;
}

i.fas.fa-times {
    cursor: pointer;
    position: relative;
    left: calc(100% - 2.2rem);
    bottom: 1.98rem;
    color: #686868;
}

.react-autosuggest__container {
    position: relative;
}
  
.react-autosuggest__input:focus {
    outline: none;
}
  
.react-autosuggest__container--open .react-autosuggest__input {
    border-bottom-left-radius: 8;
    border-bottom-right-radius: 8;
}
  
.react-autosuggest__suggestions-container {
    display: none;
}
  
.react-autosuggest__container--open .react-autosuggest__suggestions-container {
    display: block;
    position: absolute;
    top: 41px;
    width: calc(100% - 10px);
    /*border: 1px solid #aaa;*/
    background-color: #fff;
    /*font-size: 16px;*/
    border-bottom-left-radius: 8px;
    border-bottom-right-radius: 8px;
    z-index: 2;
    box-shadow: 0px 4px 8px 4px rgba(224, 228, 229, 0.35);
    font-size: 14px;
}
  
.react-autosuggest__suggestions-list {
    margin: 0;
    padding: 0;
    list-style-type: none;
}
  
.react-autosuggest__suggestion {
    cursor: pointer;
    padding: 10px 15px;
}

.react-autosuggest__suggestion .highlight {
    font-weight: 600;
}
  
.react-autosuggest__suggestion--highlighted {
    background-color: #1C76D5;
    color: #fff;
}