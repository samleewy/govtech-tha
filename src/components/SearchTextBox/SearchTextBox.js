import React, { useEffect, useRef, useState } from 'react'
import Autosuggest from 'react-autosuggest'
import { getSuggestionsApi } from '../../utils/data'
import { getHighlightIndices, highlightText } from '../../utils/highlight'
import './SearchTextBox.css'

const SearchTextBox = ({ searchQuery, setSearchQuery, onSearchClick }) => {

    const [suggestions, setSuggestions] = useState([])
    const inputRef = useRef(null)

    useEffect(() => {
        // provide initial focus onload
        inputRef.current.focus();
    }, []);

    const getSuggestionValue = suggestion => {
        // console.log('get suggestion value', suggestion)
        return suggestion
    }

    const onSuggestionsFetchRequested = async ({ value }) => {
        // console.log('on suggestion fetch', value)
        try {
            const data = await getSuggestionsApi(value);
            setSuggestions(data.suggestions)
        } catch (err) {
            console.log(err)
        }
    }

    const onSuggestionsClearRequested = () => {
        // console.log('on suggestion clear')
        setSuggestions([]);
    }

    const onSuggestionSelected = () => {
        onSearchClick();
    }

    const renderSuggestion = (suggestion) => {
        // 'child' is hard-coded to make sure it gets highlighted, change 'child' with 
        // searchQuery, for proper implementation
        const highlightIndices = getHighlightIndices(suggestion, 'child')
        return highlightText(highlightIndices, suggestion)
    }

    const shouldRenderSuggestions = (value) => {
        return value.trim().length > 2;
    }

    const onTextChange = (_, { newValue }) => {
        // console.log('onTextChange', newValue)
        setSearchQuery(newValue)
    }

    const onKeyDown = (e) => {
        if (e.key === 'Enter') onSearchClick();
    }

    const onCancelClick = () => {
        setSearchQuery("")
        onSuggestionsClearRequested()
        inputRef.current.focus()
    }

    return (
        <>
            <Autosuggest 
                suggestions={suggestions}
                onSuggestionsFetchRequested={onSuggestionsFetchRequested}
                onSuggestionsClearRequested={onSuggestionsClearRequested}
                onSuggestionSelected={onSuggestionSelected}
                getSuggestionValue={getSuggestionValue}
                renderSuggestion={renderSuggestion}
                shouldRenderSuggestions={shouldRenderSuggestions}
                inputProps={{
                    value: searchQuery,
                    onChange: onTextChange,
                    onKeyDown: onKeyDown,
                    ref: inputRef,
                    'data-testid': 'search-textbox'
                }} 
                containerProps={{
                    'data-testid': 'search-textbox-suggest'
                }}
            />
            { searchQuery && searchQuery.length >= 1 && 
            <i 
                data-testid='search-cancel-btn'
                className="fas fa-times"
                onClick={onCancelClick}
            ></i>}
        </>
    )
}

export default SearchTextBox