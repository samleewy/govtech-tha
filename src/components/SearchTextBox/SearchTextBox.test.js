import { render } from "@testing-library/react";
import SearchTextBox from "./SearchTextBox";

describe("This suite is to test SearchTextBox component", () => {
    it('should render correctly', () => {
        const { asFragment } = render(<SearchTextBox searchQuery=""/>)
        expect(asFragment()).toMatchSnapshot();
    })
    it('should not show cancel/cross button when searchQuery < 1', () => {
        const { queryByTestId } = render(<SearchTextBox searchQuery=""/>)
        const cancelBtn = queryByTestId('search-cancel-btn');
        expect(cancelBtn).toBeNull();
    })
    it('should show cancel/cross button when searchQuery >= 1', () => {
        const { getByTestId } = render(<SearchTextBox searchQuery="test"/>)
        const cancelBtn = getByTestId('search-cancel-btn');
        expect(cancelBtn).toBeInTheDocument();
    })
})