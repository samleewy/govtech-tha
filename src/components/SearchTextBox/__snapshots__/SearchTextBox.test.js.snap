// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`This suit is to test SearchTextBox component should render correctly 1`] = `
<DocumentFragment>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    class="react-autosuggest__container"
    data-testid="search-textbox-suggest"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="react-autowhatever-1"
      autocomplete="off"
      class="react-autosuggest__input react-autosuggest__input--focused"
      data-testid="search-textbox"
      type="text"
      value=""
    />
    <div
      class="react-autosuggest__suggestions-container"
      id="react-autowhatever-1"
      role="listbox"
    />
  </div>
</DocumentFragment>
`;
